relay-general/src/protocol/contexts/mod.rs

```diff
@@ -87,6 +88,19 @@ impl Context {
     }
 }
 
+/// Trait to get the Context both from the user agent string and also the new client hints.
+/// With an automatically derived function which tries to first get the context from client hints,
+/// if that fails it tries for the user agent string.
+pub trait ContextFromUserAgentInfo: Sized {
```

Comment:
nit: This trait could be implemented for things other than Contexts, so no need to put `Context` in the name.

```suggestion
pub trait FromUserAgentInfo: Sized {
```